File: Day Eight 27-10-24/solution-1.py
from typing import List

class Solution:
    def removeDuplicates(self, nums: List[int]) -> int:
        """
        Removes duplicates from a sorted list in-place and returns the number of unique elements.
        
        Args:
        nums (List[int]): A list of sorted integers where duplicates may exist.
        
        Returns:
        int: The number of unique elements in the list after removing duplicates.
        """
        # Check if the list is empty
        if not nums:
            return 0  # Return 0 if there are no elements in the list
        
        # Initialize left to 0, which will track the position of the next unique element
        left = 0
        # Iterate over the list starting from the second element
        for right in range(1, len(nums)):
            # Compare the current element with the last unique element
            if nums[left] != nums[right]:
                # If they are different, increment left and assign the current element to the position left
                left += 1
                nums[left] = nums[right]
        
        # Return the number of unique elements found, which is left + 1
        return left + 1

# Example usage of the Solution class to remove duplicates
nums = [1,1,2,3,3,4,5]
print(Solution().removeDuplicates(nums))  # Output should be 5, as there are 5 unique elements

File: Day Eight 27-10-24/solution-1.js
/*
Problem: Remove Duplicates from Sorted Array

Given an integer array nums sorted in non-decreasing order, remove the duplicates in-place such that each unique element appears only once. The relative order of the elements should be kept the same. Then return the number of unique elements in nums.

Consider the number of unique elements of nums to be k, to get accepted, you need to do the following things:

Change the array nums such that the first k elements of nums contain the unique elements in the order they were present in nums initially. The remaining elements of nums are not important as well as the size of nums.
Return k.
Custom Judge:

The judge will test your solution with the following code:

int[] nums = [...]; // Input array
int[] expectedNums = [...]; // The expected answer with correct length

int k = removeDuplicates(nums); // Calls your implementation

assert k == expectedNums.length;
for (int i = 0; i < k; i++) {
    assert nums[i] == expectedNums[i];
}
If all assertions pass, then your solution will be accepted.

Constraints:

1 <= nums.length <= 3 * 10^4
-100 <= nums[i] <= 100
nums is sorted in non-decreasing order.
*/

// Function to remove duplicates from a sorted array
const removeDuplicates = (nums) => { 
    // If the array is empty, return 0 as there are no elements
    if (nums.length === 0) return 0;

    // Initialize k to 1, which will track the position of the next unique element
    let k = 1;
    // Start iterating from the second element (index 1)
    for (let i = 1; i < nums.length; i++){
        // If the current element is not equal to the previous element
        if (nums[i] != nums[i - 1]) {
            // Assign the current element to the position k
            nums[k] = nums[i];
            // Increment k to move to the next position for a unique element
            k++;
        }
    }
    // Return the number of unique elements
    return k;
};

// Test the function with an example array
console.log(removeDuplicates([1, 1, 2, 3, 4, 4, 5])); // Output 5
